stablehlo/dialect/TypeInference.cpp

```diff
@@ -2670,10 +2670,21 @@ LogicalResult inferSelectOp(
                                "requires the same shape for all operands");
 
   // The output shape should be derived from the most specific parts of the
-  // `onTrue` and `onFalse` (see documentation for details).
-  SmallVector<Type> inferredReturnTypes;
-  return inferMostSpecificTypeComponents(location, {trueType, falseType},
-                                         inferredReturnShapes);
+  // `onTrue` and `onFalse` when the bounds are missing (see documantation for
+  // `inferMostSpecificTypeComponents` or `inferMostSpecificType` for details).
+  auto onTrueRankedTy = onTrue.getType().dyn_cast<RankedTensorType>();
+  auto onFalseRankedTy = onFalse.getType().dyn_cast<RankedTensorType>();
+  if (!onTrueRankedTy || !onFalseRankedTy)
+    return inferMostSpecificTypeComponents(location, {trueType, falseType},
+                                           inferredReturnShapes);
+  auto inferredTypeOrErr =
+      inferMostSpecificType(location, {onTrueRankedTy, onFalseRankedTy});
+  if (failed(inferredTypeOrErr)) return failure();
+  auto resultType = (*inferredTypeOrErr).cast<RankedTensorType>();
+  inferredReturnShapes.emplace_back(resultType.getShape(),
+                                    resultType.getElementType(),
+                                    resultType.getEncoding());
+  return success();
```

Comment:
Thanks for the update. How about we simplify all the new code to:
```
auto inferredTypeOrErr =
      inferMostSpecificType(location, {trueType, falseType});
  if (failed(inferredTypeOrErr)) return failure();
  auto resultType = (*inferredTypeOrErr).cast<RankedTensorType>();
  inferredReturnShapes.emplace_back(resultType.getShape(),
                                    resultType.getElementType(),
                                    resultType.getEncoding());
  return success();
```
So that we don't need to call twice, one with `inferMostSpecificTypeComponents` and another one with `inferMostSpecificType()`. Because it can handle unranked cases already.

Reply:
Why do we need to make any changes to the code at all? I think that `inferMostSpecificTypeComponents` [already supports bounds](https://github.com/openxla/stablehlo/blob/b8f97cf15495617e8fbaea98b2f3977df66ecfa8/stablehlo/dialect/Base.cpp#L303-L310)?

Reply:
I also remember for a few ops we switched from using `inferReturnTypeComponents` to `inferredReturnTypes` but switched back because `inferReturnTypeComponents` now added support for encodings.

`inferSelectOp` passes `inferredReturnShapes` as a parameter (other ops use `inferReturnTypes` which uses `inferredReturnTypes` but this op passes `inferReturnTypeComponents`), so the encoding has to be explicitly passed, but `inferMostSpecificTypeComponents` only passes the shape.
https://github.com/openxla/stablehlo/blob/main/stablehlo/dialect/Base.cpp#L363-L370

Reply:
Let's fix the linked code in Base.cpp to also pass the encoding to ShapedTypeComponents, then we'll be able to avoid code changes in TypeInference.cpp? I.e. replace `inferredReturnShapes.emplace_back(*inferredTypeOrErr);` with something like:
```C++
  inferredReturnShapes.emplace_back(resultType.getShape(),
                                    resultType.getElementType(),
                                    resultType.getEncoding());
```

Reply:
SGTM :)